Add notice and notification in Advanced Fraud Settings page

daquinons opened the pull request:
Fixes #6497 

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->
Advanced fraud protection should not be enabled if no rules are enabled. Right now, because there is no check when that happens, we save it gracefully but then it's reverted back.

This PR adds a new notice at the top of the page to indicate that at least one rule needs to be enabled, and dispatches an error notice when that's not the case and the user saves its settings.

In order to avoid weird states (such as the user enables Advanced with one rule set, saves it, but then goes back to it and disables the rule set), the level is stored as "basic" when clicking on save without any rule enabled (if the current level is not already "basic").
<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->
![image](https://github.com/Automattic/woocommerce-payments/assets/2612426/600a9ccc-ffba-49f4-8dcf-8e67d9fef4f2)


![image](https://github.com/Automattic/woocommerce-payments/assets/2612426/73c62e01-22a3-4c06-aedb-b062ab0c8c60)

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to Payments - Settings
* Browse to the bottom of the page and see if "Basic" is selected as fraud protection
* If it's not selected, select "Basic" and save
* Select "Advanced" now, and click on "Configure"
* Without enabling any rule, click on "Save"
* You'll see the notification "At least one risk filter needs to be enabled for advanced protection."
* Go back to WooPayments settings
* Check that "Basic" is still selected
* Choose "Advanced" again and click on "Configure"
* Enable any number of rulesets (but at least one)
* Click on "Save"
* You shouldn't see the error notification now, but a notification saying "advanced" is set
* Go back to WooPayments Settings and check "Advanced" is selected

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-7.6.0#advanced-fraud-protection-cant-be-enabled-without-enabling-advanced-fraud-rules
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

daquinons committed:
Add notice and notification in advanced fraud settings page

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8762` or branch name `fix/6497-advanced-fraud-settings-empty` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/6497-advanced-fraud-settings-empty) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 9469a2000ccafe1b52bf488bb793fce8da246ca6
- Build time: 2024-05-03 15:47:32 UTC

_Note: the build is updated when a new commit is pushed to this PR._

daquinons committed:
Update snapshots

daquinons committed:
Add test

daquinons committed:
Remove unused variables

daquinons committed:
Add changelog

daquinons committed:
Improvement

daquinons committed:
Merge branch 'develop' into fix/6497-advanced-fraud-settings-empty

daquinons committed:
Revert to basic

daquinons committed:
Merge branch 'fix/6497-advanced-fraud-settings-empty' of https://github.com/Automattic/woocommerce-payments into fix/6497-advanced-fraud-settings-empty

eduardoumpierre reviewed (commented):
I see 3 paths in this flow:
1. The user is currently on basic and saves an empty advanced setting.
    - In this case, couldn't we return early to avoid the "Settings saved" toast?
    - Since there are no changes to save, I think it would be safe to just skip the save and present just the `At least one risk filter needs to be enabled for advanced protection.` message.
2. The user is currently on basic and saves a filled advanced setting.
    - The `Current protection level is set to "advanced".` makes sense.
3. The user is currently on advanced setting and saves an empty advanced setting (resets the form)
    - The user will be notified by an `At least one risk filter needs to be enabled for advanced protection.` message, but I think a better message could be `Protection level set to "Basic" due to empty advanced configuration.` in this case.

What do you think?

daquinons committed:
Exit early

daquinons committed:
Modify notification message

daquinons commented:
Thanks for the suggestions @eduardoumpierre !

> I see 3 paths in this flow:
> 
> 1. The user is currently on basic and saves an empty advanced setting.
>    
>    * In this case, couldn't we return early to avoid the "Settings saved" toast?
>    * Since there are no changes to save, I think it would be safe to just skip the save and present just the `At least one risk filter needs to be enabled for advanced protection.` message.

At the beginning I assumed that we were sharing the context with the main Settings screen so that's why it had to be saved even entering on the Advanced risk screen. However, I just tested it and we are not sharing it so exiting early makes sense here :)  

There is a problem though, exiting early without saving the settings will break it in the example I give in the description: a user with advanced settings enabled and one rule enable, if it disables the rule, I wouldn't be able to move it down to "Basic" level. I've added the save only inside that check ( `if ( ProtectionLevel.BASIC !== currentProtectionLevel ) `) let me know what you think https://github.com/Automattic/woocommerce-payments/pull/8762/commits/f03083036d1d06c40b0d32f1b98160818abf7d1f

> 2. The user is currently on basic and saves a filled advanced setting.
>    
>    * The `Current protection level is set to "advanced".` makes sense.

👍 

> 3. The user is currently on advanced setting and saves an empty advanced setting (resets the form)
>    
>    * The user will be notified by an `At least one risk filter needs to be enabled for advanced protection.` message, but I think a better message could be `Protection level set to "Basic" due to empty advanced configuration.` in this case.
> 
> What do you think?

I've changed to `'Protection level set to "basic", at least one risk filter needs to be enabled for advanced protection.'` to keep the same message that Design gave us but adding the "basic" part too, let me know if you think should be in the other way around but wanted to keep what Design gave just in case :D https://github.com/Automattic/woocommerce-payments/pull/8762/commits/38ea9bceaf2062071da76746bc0a9382b312cd14

eduardoumpierre commented on client/settings/fraud-protection/advanced-settings/index.tsx:
```diff
@@ -154,9 +162,31 @@ const FraudProtectionAdvancedSettingsPage: React.FC = () => {
+					)
+				);
+
+				return;
```
I think the `wcpaySettings?.accountStatus?.fraudProtection?.declineOnAVSFailure` bit below is important because it sets the `declineOnAVSFailure ` value on the global `wcpaySettings ` object. Because of that, it would be good to refactor the logic a bit to only use the early return when we don't want to perform any updates.

This is my suggestion – I also changed the order of the function to remove the `else` from the main `if` statement so that we can remove an `if` level.

```diff
diff --git a/client/settings/fraud-protection/advanced-settings/index.tsx b/client/settings/fraud-protection/advanced-settings/index.tsx
index 36ed53fb2..f18b70a62 100644
--- a/client/settings/fraud-protection/advanced-settings/index.tsx
+++ b/client/settings/fraud-protection/advanced-settings/index.tsx
@@ -169,55 +169,59 @@ const FraudProtectionAdvancedSettingsPage: React.FC = () => {
 	};
 
 	const handleSaveSettings = () => {
-		if ( validateSettings( protectionSettingsUI ) ) {
-			const settings = writeRuleset( protectionSettingsUI );
-
-			if ( ! checkAnyRuleFilterEnabled( protectionSettingsUI ) ) {
-				if ( ProtectionLevel.BASIC !== currentProtectionLevel ) {
-					updateProtectionLevel( ProtectionLevel.BASIC );
-					updateAdvancedFraudProtectionSettings( settings );
-					saveSettings();
-				}
+		if ( ! validateSettings( protectionSettingsUI ) ) {
+			window.scrollTo( {
+				top: 0,
+			} );
+			return;
+		}
+
+		if ( ! checkAnyRuleFilterEnabled( protectionSettingsUI ) ) {
+			if ( ProtectionLevel.BASIC === currentProtectionLevel ) {
 				dispatch( 'core/notices' ).createErrorNotice(
 					__(
-						'Protection level set to "basic", at least one risk filter needs to be enabled for advanced protection.',
+						'At least one risk filter needs to be enabled for advanced protection.',
 						'woocommerce-payments'
 					)
 				);
-
 				return;
-			} else if ( ProtectionLevel.ADVANCED !== currentProtectionLevel ) {
-				updateProtectionLevel( ProtectionLevel.ADVANCED );
-				dispatch( 'core/notices' ).createSuccessNotice(
-					__(
-						'Current protection level is set to "advanced".',
-						'woocommerce-payments'
-					)
-				);
-			}
-
-			// Persist the AVS verification setting until the account cache is updated locally.
-			if (
-				wcpaySettings?.accountStatus?.fraudProtection
-					?.declineOnAVSFailure
-			) {
-				wcpaySettings.accountStatus.fraudProtection.declineOnAVSFailure = settings.some(
-					( setting ) => setting.key === 'avs_verification'
-				);
 			}
 
-			updateAdvancedFraudProtectionSettings( settings );
+			updateProtectionLevel( ProtectionLevel.BASIC );
+			dispatch( 'core/notices' ).createErrorNotice(
+				__(
+					'Current protection level is set to "basic". At least one risk filter needs to be enabled for advanced protection.',
+					'woocommerce-payments'
+				)
+			);
+		} else if ( ProtectionLevel.ADVANCED !== currentProtectionLevel ) {
+			updateProtectionLevel( ProtectionLevel.ADVANCED );
+			dispatch( 'core/notices' ).createSuccessNotice(
+				__(
+					'Current protection level is set to "advanced".',
+					'woocommerce-payments'
+				)
+			);
+		}
 
-			saveSettings();
+		const settings = writeRuleset( protectionSettingsUI );
 
-			recordEvent( 'wcpay_fraud_protection_advanced_settings_saved', {
-				settings: JSON.stringify( settings ),
-			} );
-		} else {
-			window.scrollTo( {
-				top: 0,
-			} );
+		// Persist the AVS verification setting until the account cache is updated locally.
+		if (
+			wcpaySettings?.accountStatus?.fraudProtection?.declineOnAVSFailure
+		) {
+			wcpaySettings.accountStatus.fraudProtection.declineOnAVSFailure = settings.some(
+				( setting ) => setting.key === 'avs_verification'
+			);
 		}
+
+		updateAdvancedFraudProtectionSettings( settings );
+
+		saveSettings();
+
+		recordEvent( 'wcpay_fraud_protection_advanced_settings_saved', {
+			settings: JSON.stringify( settings ),
+		} );
 	};
 
 	// Hack to make "Payments > Settings" the active selected menu item.

```

Let me know what you think about it!

daquinons replied:
Thanks a lot, Eduardo, the changes make it much better structured, I overcomplicated it 😵

Merged https://github.com/Automattic/woocommerce-payments/pull/8762/commits/9469a2000ccafe1b52bf488bb793fce8da246ca6

daquinons committed:
Refactor

eduardoumpierre reviewed (approved):
I followed the testing instructions and it worked as expected! The code is clear and covered by tests.

Thanks for working on it! LGTM 🚀
